### PR TITLE
2959 email preview

### LINKED ISSUE #2959: Fix email preview error on invalid ID
Email previews can be passed a user id as a get parameter `id=`. If the id doesn't exist or is the wrong type of user for the email, don't throw an unhandled error, instead show an error explaining why the id parameter is bad.

see https://casa-qa.herokuapp.com/rails/mailers
see `lib/mailers/previews/`

### PR DESCRIPTION
**Pull request by elhalvers**
### What github issue is this PR for, if any?
Resolves #2959 

### What changed, and why?
1. We created a DebugPreviewMailer class that displays information to the developer in the event that an invalid user_id gets passed during email previews. We believe that this may be the best approach as the mailer previews do not have access to methods like `render` or `redirect_to` to display information. Email previews expect to render an email so we created a debugging mail preview to render in the case of an invalid id.
2. This solution will not throw errors when an invalid id is passed to a preview mailer.
3. Code added to the `sent_email_event` initializer to skip the logging and `SentEmail` creation when the DebugPreviewMailer is called. This avoids errors that would normally occur when a nil user is provided. 

### How will this affect user permissions?
NA

### How is this tested? (please write tests!) 💖💪
Modified existing MailerPreview tests to include scenarios where no id is passed, a valid id is passed or an invalid id is passed.
When an invalid id is passed, the tests check that our DebugPreviewMailer solution is working as expected.

### Screenshots please :)

Screenshot 2023-04-17 at 2 40 37 PM

### Feelings gif (optional)
What gif best describes your feeling working on this issue? https://giphy.com/
How to embed:
`alt text`

### Feedback please? (optional)
We are very interested in your feedback! Please give us some :) https://forms.gle/1D5ACNgTs2u9gSdh9

### CODE REVIEW

File: lib/mailers/debug_preview_mailer.rb [hunk 1]
```diff
@@ -0,0 +1,10 @@
+class DebugPreviewMailer < ActionMailer::Base
+  def invalid_user(role)
+    mail(
+      from: "reply@example.com",
+      to: "missing_#{role}@example.com",
+      subject: "invalid_user_id",
+      body: "User does not exist or is not a #{role}"
+    )
+  end
+end
```
Comment (cnorm35) [inline lib/mailers/debug_preview_mailer.rb:10]: I like the solution you came up with here! Keeps things clean and re-usable across the other mailers 🏅

**Comment by cnorm35** [comment]
Looking good!  Once the codeclimate CI check is back to passing, everything should be good for approval

https://codeclimate.com/github/rubyforgood/casa/pull/4767

**Comment by elhalvers** [comment]
Cool! Thanks!Josh came up with the idea and I liked it as well!: )EricOn Apr 18, 2023, at 09:03, Cody Norman ***@***.***> wrote:﻿
@cnorm35 commented on this pull request.



In lib/mailers/debug_preview_mailer.rb:
> @@ -0,0 +1,10 @@
+class DebugPreviewMailer < ActionMailer::Base
+  def invalid_user(role)
+    mail(
+      from: ***@***.***",
+      to: ***@***.***",
+      subject: "invalid_user_id",
+      body: "User does not exist or is not a #{role}"
+    )
+  end
+end

I like the solution you came up with here! Keeps things clean and re-usable across the other mailers 🏅

**Comment by elhalvers** [comment]
Awesome! Anything we need to do on our end for this?Thanks!: )EricOn Apr 18, 2023, at 09:12, Cody Norman ***@***.***> wrote:﻿
Looking good!  Once the codeclimate CI check is back to passing, everything should be good for approval
https://codeclimate.com/github/rubyforgood/casa/pull/4767

**Review by dpaola2 — APPROVED** [review]
This is good work. Wish I had more feedback!

**Comment by dpaola2** [comment]
Take a look at the code climate issues

**Comment by elhalvers** [comment]
Thanks Dave!

: )

**Comment by JoshDevHub** [comment]
I came up with this idea to potentially DRY it up some:
```rb
# lib/mailers/previews/application_mailer_preview.rb

# new class for other previews to inherit from
class ApplicationMailerPreview < ActionMailer::Preview
  # the method that creates the preview and calls its method https://github.com/rails/rails/blob/main/actionmailer/lib/action_mailer/preview.rb#L112-L117
  def self.call(...)
    super(...)
  rescue ActiveRecord::RecordNotFound
    DebugPreviewMailer.invalid_user(
      name.delete_suffix("MailerPreview").underscore
    )
  end
end

# and then a preview mailer method could be like ->
class CasaAdminMailerPreview < ApplicationMailerPreview
  def account_setup
    # using `::find` so that `ActiveRecord::RecordNotFound` is raised on failed lookup
    casa_admin = params.key?(:id) ? CasaAdmin.find(params[:id]) : CasaAdmin.last
  
    CasaAdminMailer.account_setup(casa_admin)
  end
end

**Comment by elhalvers** [comment]
Cool : )

I like your `class ApplicationMailerPreview < ActionMailer::Preview`
pattern : )

**Review by cnorm35 — APPROVED** [review]

**Pull request merged**